File: modules/headphones/android/src/main/java/expo/modules/headphones/HeadphonesModule.kt
package expo.modules.headphones

import android.content.Context
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Detecta si hay auriculares (cable, USB o Bluetooth) conectados a la salida de audio.
 * PostureFix lo usa para decidir si reproduce el tono EAS directamente en los oídos
 * del usuario en lugar de por el altavoz.
 */
class HeadphonesModule : Module() {
  private val audioManager: AudioManager
    get() {
      val context = requireNotNull(appContext.reactContext) { "React context is not available" }
      return context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

  private var deviceCallback: AudioDeviceCallback? = null

  override fun definition() = ModuleDefinition {
    Name("Headphones")

    Events("onChange")

    Function("isConnected") {
      currentOutput().first
    }

    Function("getStatus") {
      val (connected, kind) = currentOutput()
      mapOf("connected" to connected, "kind" to kind)
    }

    OnStartObserving {
      val callback = object : AudioDeviceCallback() {
        override fun onAudioDevicesAdded(addedDevices: Array<out AudioDeviceInfo>?) = emitStatus()
        override fun onAudioDevicesRemoved(removedDevices: Array<out AudioDeviceInfo>?) = emitStatus()
      }
      deviceCallback = callback
      audioManager.registerAudioDeviceCallback(callback, null)
    }

    OnStopObserving {
      deviceCallback?.let { audioManager.unregisterAudioDeviceCallback(it) }
      deviceCallback = null
    }

    OnDestroy {
      deviceCallback?.let { audioManager.unregisterAudioDeviceCallback(it) }
      deviceCallback = null
    }
  }

  private fun emitStatus() {
    val (connected, kind) = currentOutput()
    sendEvent("onChange", mapOf("connected" to connected, "kind" to kind))
  }

  /** @return `true` y el tipo de salida cuando el audio sale por unos auriculares. */
  private fun currentOutput(): Pair<Boolean, String?> {
    val devices = audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
    for (device in devices) {
      val kind = when (device.type) {
        AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
        AudioDeviceInfo.TYPE_WIRED_HEADSET -> "wired"
        AudioDeviceInfo.TYPE_USB_HEADSET,
        AudioDeviceInfo.TYPE_USB_DEVICE -> "usb"
        AudioDeviceInfo.TYPE_BLUETOOTH_A2DP,
        AudioDeviceInfo.TYPE_BLUETOOTH_SCO -> "bluetooth"
        AudioDeviceInfo.TYPE_HEARING_AID -> "hearing-aid"
        else -> null
      }
      if (kind != null) {
        return true to kind
      }
    }
    return false to null
  }
}

File: modules/headphones/ios/HeadphonesModule.swift
import AVFoundation
import ExpoModulesCore

/**
 Detecta si la ruta de salida de audio actual son unos auriculares (cable, USB,
 Bluetooth o Lightning). PostureFix lo usa para lanzar el tono EAS directamente
 a los oídos del usuario en lugar de por el altavoz.
 */
public class HeadphonesModule: Module {
  private var observer: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("Headphones")

    Events("onChange")

    Function("isConnected") {
      currentOutput().connected
    }

    Function("getStatus") {
      let output = currentOutput()
      return ["connected": output.connected, "kind": output.kind as Any]
    }

    OnStartObserving {
      self.observer = NotificationCenter.default.addObserver(
        forName: AVAudioSession.routeChangeNotification,
        object: AVAudioSession.sharedInstance(),
        queue: .main
      ) { [weak self] _ in
        guard let self else { return }
        let output = self.currentOutput()
        self.sendEvent("onChange", ["connected": output.connected, "kind": output.kind as Any])
      }
    }

    OnStopObserving {
      self.removeObserver()
    }

    OnDestroy {
      self.removeObserver()
    }
  }

  private func removeObserver() {
    if let observer {
      NotificationCenter.default.removeObserver(observer)
    }
    observer = nil
  }

  private func currentOutput() -> (connected: Bool, kind: String?) {
    for output in AVAudioSession.sharedInstance().currentRoute.outputs {
      switch output.portType {
      case .headphones, .headsetMic:
        return (true, "wired")
      case .usbAudio:
        return (true, "usb")
      case .bluetoothA2DP, .bluetoothHFP, .bluetoothLE:
        return (true, "bluetooth")
      default:
        continue
      }
    }
    return (false, nil)
  }
}
